internal/pgadmin/users.go

```diff
@@ -57,103 +75,100 @@ spec = importlib.util.find_spec('.pgadmin', (
-
-	// Due to limitations on the types of updates that can be made to active server
-	// connections, when the current server connection is updated, we need to delete
-	// it and add a new server connection in its place. This will require a refresh
-	// if pgAdmin web GUI is being used when the update takes place.
-	// - https://git.postgresql.org/gitweb/?p=pgadmin4.git;f=web/pgadmin/browser/server_groups/servers/__init__.py;hb=REL-4_20#l604
-
-	// define the hostname of the primary service
-	primary := naming.ClusterPrimaryService(cluster)
-	hostname := primary.Name + "." + primary.Namespace + ".svc"
-
-	const script = `
+    sys.path.insert(0, root)`,
+
+		// Import pgAdmin modules now that they are on the search path.
+		// NOTE: When testing with the REPL, use the `__enter__` method to
+		// avoid one level of indentation.
+		//
+		//     create_app().app_context().__enter__()
+		//
+		`
 import copy
 import json
 import sys
+
 from pgadmin import create_app
 from pgadmin.model import db, Role, User, Server, ServerGroup
+from pgadmin.utils.constants import INTERNAL
 from pgadmin.utils.crypto import encrypt
 
-with create_app().app_context():
+with create_app().app_context():`,
+
+		// The user with id=1 is automatically created by pgAdmin when it
+		// creates its configuration database. Clear that email and username
+		// so they cannot conflict with users we create, and deactivate the user
+		// so it cannot log in.
+		// - https://git.postgresql.org/gitweb/?p=pgadmin4.git;f=web/migrations/versions/fdc58d9bd449_.py;hb=REL-4_30#l129
+		`
     admin = db.session.query(User).filter_by(id=1).first()
     admin.active = False
     admin.email = ''
     admin.password = ''
+    admin.username = ''
 
     db.session.add(admin)
-    db.session.commit()
-
+    db.session.commit()`,
+
+		// Process each line of input as a single user definition. Those with
+		// a non-blank password are allowed to login.
+		//
+		// The "internal" authentication source requires that username and email
+		// be the same and be an email address. Append "@pgo" to the username
+		// to pass login validation.
+		// - https://git.postgresql.org/gitweb/?p=pgadmin4.git;f=web/pgadmin/authenticate/internal.py;hb=REL-4_30#l88
+		// - https://git.postgresql.org/gitweb/?p=pgadmin4.git;f=web/pgadmin/utils/validation_utils.py;hb=REL-4_30#l13
+		//
+		// The "auth_source" and "username" attributes are part of the User
+		// model since pgAdmin v4.21.
+		// - https://git.postgresql.org/gitweb/?p=pgadmin4.git;f=web/pgadmin/model/__init__.py;hb=REL-4_30#l66
+		`
     for line in sys.stdin:
         if not line.strip():
             continue
 
         data = json.loads(line)
+        address = data['username'] + '@pgo'
         user = (
-            db.session.query(User).filter_by(email=data['username']).first() or
-            User(email=data['username'])
+            db.session.query(User).filter_by(username=address).first() or
+            User()
```

Comment:
❓ would the user ever put in an email address on their own? just wondering if we need to check for the @ first before just padding it with @pgo

Reply:
I hadn't considered that. We create pgAdmin users from the list of Postgres users in `spec.users` where we don't allow `@`: (I forgot about this and tried it out.)

```
The PostgresCluster "mycluster" is invalid: spec.users.name: Invalid value: "human@computer": spec.users.name in body should match '^[a-z0-9]([-a-z0-9]*[a-z0-9])?$'
```

I tried it in the guts pgAdmin anyway and found that we can write it, but the user isn't able to log in. They get an error: `Email/Username is not valid`

```python
>>> user = User()
>>> user.auth_source = INTERNAL
>>> user.email = user.username = 'human@some@where'
>>> user.password = 'passwordpassword'
>>> user.active = True
>>> user.roles = db.session.query(Role).filter_by(name='User').all()
>>> user.masterpass_check = 'any'
>>> user.verify_and_update_password(user.password)
True
>>> db.session.add(user)
>>> db.session.commit()
```

So, for now, at least, I think we can ignore it. Validation blocks it, and bypassing validation won't crash PGO or block reconciles.

Reply:
oh cool, it's caught in the spec 👍